File: android-tv/app/src/main/java/com/donghuaflix/data/repository/ShowRepository.kt
package com.donghuaflix.data.repository

import com.donghuaflix.data.local.dao.ShowDao
import com.donghuaflix.data.local.dao.WatchHistoryDao
import com.donghuaflix.data.mapper.ShowMapper
import com.donghuaflix.data.remote.DonghuaApi
import com.donghuaflix.domain.model.Episode
import com.donghuaflix.domain.model.Show
import com.donghuaflix.domain.model.SubtitleTrack
import com.donghuaflix.domain.model.VideoSource
import kotlinx.coroutines.flow.Flow
import kotlinx.coroutines.flow.map
import javax.inject.Inject
import javax.inject.Singleton

@Singleton
class ShowRepository @Inject constructor(
    private val api: DonghuaApi,
    private val showDao: ShowDao,
    private val watchHistoryDao: WatchHistoryDao,
) {
    fun observeAllShows(): Flow<List<Show>> = showDao.getAllShows().map { entities ->
        entities.map { ShowMapper.entityToDomain(it) }
    }

    suspend fun getShowsFromApi(website: String, pageSize: Int = 20): List<Show> {
        return try {
            val result = api.getShows(page = 1, pageSize = pageSize, website = website)
            result.items.map { ShowMapper.dtoToDomain(it) }
        } catch (e: Exception) {
            emptyList()
        }
    }

    suspend fun getShow(id: Int): Show? {
        // Try local first
        val local = showDao.getShowById(id)
        if (local != null) {
            val progress = watchHistoryDao.getLastWatchedForShow(id)
            return ShowMapper.entityToDomain(local, progress?.let { ShowMapper.watchHistoryToDomain(it) })
        }

        // Fallback to API
        return try {
            val dto = api.getShow(id)
            showDao.upsertShow(ShowMapper.dtoToEntity(dto))
            ShowMapper.dtoToDomain(dto)
        } catch (e: Exception) {
            null
        }
    }

    suspend fun searchShows(query: String): List<Show> {
        // Search locally first
        val local = showDao.searchShows(query)
        if (local.isNotEmpty()) return local.map { ShowMapper.entityToDomain(it) }

        // Fallback to API
        return try {
            val result = api.searchShows(query)
            val entities = result.items.map { ShowMapper.dtoToEntity(it) }
            showDao.upsertShows(entities)
            result.items.map { ShowMapper.dtoToDomain(it) }
        } catch (e: Exception) {
            emptyList()
        }
    }

    suspend fun getShowsByGenre(genre: String): List<Show> {
        val local = showDao.getShowsByGenre(genre)
        return local.map { ShowMapper.entityToDomain(it) }
    }

    suspend fun getRecentShows(limit: Int = 20): List<Show> {
        return showDao.getRecentShows(limit).map { ShowMapper.entityToDomain(it) }
    }

    suspend fun getCompletedShows(): List<Show> {
        return showDao.getShowsByStatus("completed").map { ShowMapper.entityToDomain(it) }
    }

    suspend fun getEpisodes(showId: Int, website: String? = null): List<Episode> {
        return try {
            api.getShowEpisodes(showId, website).map {
                Episode(it.id, it.episodeNumber, it.title, it.externalUrl, it.websiteName)
            }
        } catch (e: Exception) {
            emptyList()
        }
    }

    suspend fun getEpisodeSources(episodeId: Int): List<VideoSource> {
        return try {
            api.getEpisodeSources(episodeId).map {
                VideoSource(it.id, it.sourceName, it.sourceKey, it.sourceUrl, it.sourceType, it.websiteName)
            }
        } catch (e: Exception) {
            emptyList()
        }
    }

    suspend fun resolveSource(sourceId: Int): VideoSource? {
        return try {
            val dto = api.resolveSource(sourceId)
            val subtitles = dto.subtitles?.mapNotNull { (lang, sub) ->
                val url = sub.url ?: return@mapNotNull null
                SubtitleTrack(
                    language = lang,
                    label = sub.label ?: lang,
                    url = url,
                )
            } ?: emptyList()
            VideoSource(dto.id, dto.sourceName, dto.sourceKey, dto.sourceUrl, dto.sourceType, dto.websiteName, subtitles)
        } catch (e: Exception) {
            null
        }
    }

    suspend fun getAllGenres(): List<String> {
        val localGenres = showDao.getAllGenreStrings()
            .flatMap { it.split(",") }
            .map { it.trim() }
            .filter { it.isNotBlank() }
            .distinct()
            .sorted()

        if (localGenres.isNotEmpty()) return localGenres

        return try {
            api.getGenres()
        } catch (e: Exception) {
            emptyList()
        }
    }
}

File: apple/DonghuaFlix/Core/Repositories/ShowRepository.swift
import Foundation
import Observation

@Observable
final class ShowRepository {

    private let apiClient: APIClient

    init(apiClient: APIClient) {
        self.apiClient = apiClient
    }

    // MARK: - Shows

    func getShows(page: Int = 1, pageSize: Int = 20, genre: String? = nil, status: String? = nil, category: String? = nil, website: String? = nil) async throws -> ShowListDTO {
        try await apiClient.request(.shows(page: page, pageSize: pageSize, genre: genre, status: status, category: category, website: website))
    }

    func getShow(id: Int) async throws -> Show? {
        let dto: ShowDTO = try await apiClient.request(.show(id: id))
        return dto.toDomain()
    }

    func searchShows(query: String, page: Int = 1, pageSize: Int = 20) async throws -> [Show] {
        let dto: ShowListDTO = try await apiClient.request(.searchShows(query: query, page: page, pageSize: pageSize))
        return dto.items.map { $0.toDomain() }
    }

    // MARK: - Episodes & Sources

    func getEpisodes(showId: Int, website: String? = nil) async throws -> [Episode] {
        let dtos: [EpisodeDTO] = try await apiClient.request(.episodes(showId: showId, website: website))
        return dtos.map { $0.toDomain() }
    }

    func getEpisodeSources(episodeId: Int) async throws -> [VideoSource] {
        let dtos: [SourceDTO] = try await apiClient.request(.episodeSources(episodeId: episodeId))
        return dtos.map { $0.toDomain() }
    }

    func resolveSource(sourceId: Int) async throws -> VideoSource? {
        let dto: SourceDTO = try await apiClient.request(.resolveSource(sourceId: sourceId))
        return dto.toDomain()
    }

    // MARK: - Discovery

    func getAllGenres() async throws -> [String] {
        try await apiClient.request(.genres)
    }

    func getHomeSections() async throws -> [HomeSection] {
        let response: HomeResponseDTO = try await apiClient.request(.home)
        return response.sections.map { $0.toDomain() }
    }

    func getShowWebsites(id: Int) async throws -> [WebsiteInfo] {
        let dtos: [WebsiteInfoDTO] = try await apiClient.request(.showWebsites(id: id))
        return dtos.map { $0.toDomain() }
    }
}
